### ISSUE TITLE
Kong exposing upstream with AWS Signature Version 4 for AuthN/AuthZ

### ISSUE THREAD

**Issue by dgonzalezc** [Kong/kong#6405]
Requirement
- Expose a backend service(AWS) with AWS Signature Version 4:
  - AWS region
  - AccessKey
  - SecretKey
  - Service Name

Tests
- Generate Headers Needed(Postman)
  - Authorization
  - X-Amz-Date
- Add Headers with Request-Transformer plugin on each request
  - Not working because of AWS Headers Authorization date validation

Problem
- Kong only can add static headers/body/queryparams with Request-Transformer plugin

Setup
- Kong 0.13.1 + PostgreSQL 9.6

Questions
- Is there a way for Kong to manage this use case against AWS Signature 4?

**Comment by gszr** [comment]
Hey @dgonzalezc,

It sounds like the [serverless functions](https://docs.konghq.com/hub/kong-inc/serverless-functions/) plugin could be a good fit here. Is the issue that you need to dynamically generate the `X-Amz-Date` header? If so, something along the lines of the following might help:
```
curl -i -X POST http://localhost:8001/plugins \
     -F 'name=pre-function' \
     -F 'config.access[1]=kong.service.request.add_header("x-amz-date", os.date("!%Y%m%dT%H%M%S"))'
```

**Comment by dgonzalezc** [comment]
Thanks for the quick response @gszr 

AWS signature generate two headers with date/hour information for validation propuses:

- Authorization: With the date is generated the authorizacion credentials for AWS:
Credential=Access Key/20200929/region/service name

- X-Amz-Date: For validating against the Authorization

pre-function plugin i think is not available in version 0.13.1 so i will try to replicate this over Kong 0.14.*

**Comment by gszr** [comment]
@dgonzalezc Please use a newer Kong version! A lot has happened in Kong Land since 0.14. Our current version is 2.1, and 2.2 is [coming out soon](https://discuss.konghq.com/t/kong-2-2-0-alpha-1-is-released/7118). I'm closing this issue, but do feel free to post more questions and mention me.
